### PR TITLE
feat: Add labels on Application's k8s events (#11381)

### LINKED ISSUE #11381: Exposing Application labels in events
# Summary

I would like to have Application labels also in argo-cd events, the same way as it is possible to have them in Prometheus metrics (see https://argo-cd.readthedocs.io/en/stable/operator-manual/metrics/#exposing-application-labels-as-prometheus-metrics)

# Motivation

Our use-case is almost the same as the one stated in the documentation for [`Exposing Application labels as Prometheus metrics`](https://argo-cd.readthedocs.io/en/stable/operator-manual/metrics/#exposing-application-labels-as-prometheus-metrics).
We have several Application labels which differentiate team (business application) and environment (dev/test/prod) and we would like to forward ArgoCD Application events (as well as container stdout) to central ELK, particular index.
Unfortunately we don't have strict naming convention for Kubernetes namespaces or ArgoCD application which we could use for the corrent messages distribution.

# Proposal

The implementation should be similar to the implementation for Prometheus metrics - specify application labels for events and use them in `event.metadata.anotations` (like `dest-namespace` and `dest-server` [here](https://github.com/argoproj/argo-cd/blob/master/util/argo/audit_logger.go#L103) and [here](https://github.com/argoproj/argo-cd/blob/a259654e32ea93b1252eaad09d4422930112b85d/util/argo/audit_logger.go#L68)).

Finaly I would like to ask if there is some workaround that can we use now.

Thanks

### PR DESCRIPTION
**Pull request by svghadi**
Closes https://github.com/argoproj/argo-cd/issues/11381

This PR implements https://github.com/argoproj/argo-cd/issues/11381 enhancement to expose application labels on k8s events generated by Argo CD. 

Change:
This PR introduces two new key `resource.includeEventLabelKeys` & `resource.excludeEventLabelKeys` in `argocd-cm` configmap which takes in a comma separated string of label keys. 
For any events generated for Applications containing the label keys specified in `resource.includeEventLabelKeys`, the controller adds the matched label onto the event thus establishing an easy link between the event and the application to filter-out/process later using labels. The specified label keys are matched against the Application and it's AppProject labels. In case of conflict between labels on Application and it's AppProject, the Application label values are prioritised and added to the event. 
Furthermore, labels specified in `resource.excludeEventLabelKeys`  are removed from the event labels, if they were included.

Sample output:
```bash
$ kubectl get events -l env=dev --show-labels
LAST SEEN   TYPE     REASON               OBJECT                                      MESSAGE                                                                LABELS
25s         Normal   ResourceCreated      application/test-labels-on-app-k8s-events   admin created application                                              app=test,env=dev
25s         Normal   OperationStarted     application/test-labels-on-app-k8s-events   admin initiated sync to  (7e2013dcfdd0c801eececd38aa5f3db898e29f4b)    app=test,env=dev
24s         Normal   OperationCompleted   application/test-labels-on-app-k8s-events   Sync operation to 7e2013dcfdd0c801eececd38aa5f3db898e29f4b succeeded   app=test,env=dev
24s         Normal   ResourceUpdated      application/test-labels-on-app-k8s-events   Updated sync status:  -> Synced                                        app=test,env=dev
24s         Normal   ResourceUpdated      application/test-labels-on-app-k8s-events   Updated health status:  -> Progressing                                 app=test,env=dev
17s         Normal   ResourceUpdated      application/test-labels-on-app-k8s-events   Updated health status: Progressing -> Healthy                          app=test,env=dev
```

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: docs/operator-manual/argocd-cm.yaml [hunk 1]
```diff
@@ -221,6 +221,13 @@ data:
   # An optional comma-separated list of metadata.labels to observe in the UI.
   resource.customLabels: tier
 
+  # An optional comma-separated list of metadata.labels keys to add onto k8s events generated for Applications.
+  # The keys are compared against Application and it's AppProject, if matched,
+  # the corresponding labels are added on the generated event. 
+  # In case of conflict between labels on Application and AppProject, 
+  # the Application label values are prioritized and added to the event.
+  resource.eventLabelKeys: team,env
```
Comment (todaywasawesome) [inline docs/operator-manual/argocd-cm.yaml:229]: Wdyt about `resource.includeLabelKeys` and `resource.excludeLabelKeys` could have wildcards as an improvement - @leoluz

**Review by todaywasawesome — CHANGES_REQUESTED** [review]
I think this is a great start, to make it feature complete can we handle the include/exclude use cases?

**Comment by svghadi** [comment]
Thanks for @todaywasawesome & @leoluz for your inputs. I have incorporated your suggestions by introducing two new keys `resource.includeEventLabelKeys` & `resource.excludeEventLabelKeys` in `argocd-cm` configmap and adding wildcard support to make it feature complete. The PR is ready for review.

File: docs/operator-manual/argocd-cm.yaml (hunk 1 above)
Comment (svghadi) [reply docs/operator-manual/argocd-cm.yaml]: Added new keys `resource.includeEventLabelKeys` & `resource.excludeEventLabelKeys`.

File: controller/appcontroller.go [hunk 2]
```diff
@@ -939,7 +939,7 @@ func (ctrl *ApplicationController) processAppOperationQueueItem() (processNext b
 				Message: err.Error(),
 			})
 			message := fmt.Sprintf("Unable to delete application resources: %v", err.Error())
-			ctrl.auditLogger.LogAppEvent(app, argo.EventInfo{Reason: argo.EventReasonStatusRefreshed, Type: v1.EventTypeWarning}, message, "")
+			ctrl.logAppEvent(app, argo.EventInfo{Reason: argo.EventReasonStatusRefreshed, Type: v1.EventTypeWarning}, message, context.TODO())
```
Comment (jannfis) [inline controller/appcontroller.go:942]: Was wondering if we had a better context than the `TODO()` one here and elsewhere.

**Review by jannfis — APPROVED** [review]
LGTM. @todaywasawesome are your concerns addressed?

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Review by todaywasawesome — APPROVED** [review]
Great! Thanks for getting this over the line.

**Pull request merged**